src/lib/util/logger.js

```diff
@@ -1,35 +1,28 @@
 const log = require('pino')()
 const auditLogger = log.child({tag: 'AUDIT'})
+const { buildPath } = require('./graphqlPathUtil')
+const { getClientInfoFromHeaders } = require('./internal/loggerHelper')
 
 const auditLogEnabled = process.env.AUDIT_LOGGING !== 'false' && process.env.AUDIT_LOGGING !== false
 
 function auditLog (success, request, info, parent, args, msg) {
   if (auditLogEnabled) {
     auditLogger.info({
-      msg: msg || '',
-      requestId: request ? request.id : '',
-      operationType: info.operation.operation,
-      fieldName: info.fieldname,
-      parentTypeName: info.parentType.name,
-      path: buildPath(info.path),
-      success: success,
-      parent: parent,
-      arguments: args,
-      dataSourceType: info.dataSourceType || ''
+      audit: {
+        msg: msg || '',
+        requestId: request ? request.id : '',
+        operationType: info.operation.operation,
+        fieldName: info.fieldname,
+        parentTypeName: info.parentType.name,
+        path: buildPath(info.path),
+        success: success,
+        parent: parent,
+        arguments: args,
+        dataSourceType: info.dataSourceType || '',
+        clientInfo: getClientInfoFromHeaders(request)
```

Comment:
IMO, this is not the best way to approach this, this would mean base64 decoding and JSON parsing the header every time an audit log happens (which can happen several times per resolver). I think we should be parsing this once as soon as the request comes in.

I think a useful way to do this would be using the `context` object. The `context` object is a special object that we can put any information into. Usually, it's stuff from an incoming request object. A real world use case would be taking a cookie or a token, parsing it into a user object and mounting that at `context.user` which would then be available in all of our resolvers.

If you take a look here: https://github.com/aerogear/data-sync-server/blob/master/server/apolloServer.js#L26-L33

We're already the request object to the context. (`context.request`) as well as some auth stuff. I think the parsing should be done at the server level and I see two potential ways of doing it.

1. We modify the context function I linked above to parse the header and provide something like `context.clientInfo` (or you could parse the header and then replace its value directly in the request, such that context.request['data-sync-client-info'] would be the parsed object).
2. We add an express middleware to the GraphQL endpoint but before the executable schema middleware. This middleware would do the parsing and mount it directly on the request.headers object.

I have no strong opinion over which one we choose, although `context.clientInfo` would be pretty nice because it reduces the amount of HTTP related stuff we have inside our resolvers, which in theory, are supposed to be transport independent.

I hope that makes sense

Reply:
Good catch @darahayes .
Thanks for the suggestions. I will try to go with the middleware way

Reply:
`context` way seems more straight forward.

I will even go one step further and move the "auditLog" function into the context. Currently it is defined in the core lib but it makes more sense if we define it in the server and pass it to core.

Reply:
+1